Update algorithm to find ChromeDriver to download

JohnChen0 opened the pull request:
Find ChromeDriver to download, based on recommendation from
https://sites.google.com/a/chromium.org/chromedriver/downloads/version-selection

JohnChen0 committed:
Update algorithm of find ChromeDriver to download

Find ChromeDriver to download, based on recommendation from
https://sites.google.com/a/chromium.org/chromedriver/downloads/version-selection

Hexcles commented on tools/wpt/browser.py:
```diff
@@ -446,10 +446,26 @@ def find_binary(self, venv_path=None, channel=None):
+        if chrome_version is not None:
+            parts = chrome_version.split(".")
+            if len(parts) == 4:
+                latest_url = "http://chromedriver.storage.googleapis.com/LATEST_RELEASE_%s.%s.%s" % (
```
```suggestion
                latest_url = "https://chromedriver.storage.googleapis.com/LATEST_RELEASE_%s.%s.%s" % (
```

Hexcles commented on tools/wpt/browser.py:
```diff
@@ -446,10 +446,26 @@ def find_binary(self, venv_path=None, channel=None):
+                try:
+                    latest = get(latest_url).text.strip()
+                except:
+                    latest_url = "http://chromedriver.storage.googleapis.com/LATEST_RELEASE_%s" % parts[0]
```
```suggestion
                    latest_url = "https://chromedriver.storage.googleapis.com/LATEST_RELEASE_%s" % parts[0]
```

Hexcles commented on tools/wpt/browser.py:
```diff
@@ -446,10 +446,26 @@ def find_binary(self, venv_path=None, channel=None):
+                    except:
+                        pass
+        if latest is None:
+            latest = get("http://chromedriver.storage.googleapis.com/LATEST_RELEASE").text.strip()
```
```suggestion
            latest = get("https://chromedriver.storage.googleapis.com/LATEST_RELEASE").text.strip()
```

Hexcles commented on tools/wpt/browser.py:
```diff
@@ -446,10 +446,26 @@ def find_binary(self, venv_path=None, channel=None):
+                        pass
+        if latest is None:
+            latest = get("http://chromedriver.storage.googleapis.com/LATEST_RELEASE").text.strip()
         url = "http://chromedriver.storage.googleapis.com/%s/chromedriver_%s.zip" % (latest,
```
```suggestion
        url = "https://chromedriver.storage.googleapis.com/%s/chromedriver_%s.zip" % (latest,
```

Hexcles reviewed (commented):
Thanks for the patch, John!

Again regarding the dev channel (which is heavily used in WPT), IIUC, this code will eventually fall back to download `https://chromedriver.storage.googleapis.com/LATEST_RELEASE`, which is currently 2.46 and seems to be older than ChromeDriver 73.0.3683.20 (and probably doesn't work with Chrome Dev 74). I left some thoughts regarding the dev channel in https://crbug.com/chromedriver/2656#c13 . WDYT?

JohnChen0 commented:
> Again regarding the dev channel (which is heavily used in WPT), IIUC, this code will eventually fall back to download `https://chromedriver.storage.googleapis.com/LATEST_RELEASE`, which is currently 2.46 and seems to be older than ChromeDriver 73.0.3683.20 (and probably doesn't work with Chrome Dev 74). I left some thoughts regarding the dev channel in https://crbug.com/chromedriver/2656#c13 . WDYT?

ChromeDriver 2.46 actually contains slightly more features than ChromeDriver 73.0.3683.20, even though it was built earlier, since 2.46 was built from master branch, while 73.0.3683.20 was built from a release branch that was created earlier. That's the reason the non-versioned LATEST_RELEASE still points to 2.46.

I have implemented the suggestion in https://crbug.com/chromedriver/2656#c13. Does WPT routinely run tests against the latest Dev release? I think the right thing to do is update this PR to fallback to latest ChromeDriver at https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html. WDYT?

gsnedders commented:
> Does WPT routinely run tests against the latest Dev release?

More often than stable.

Hexcles commented:
> I think the right thing to do is update this PR to fallback to latest ChromeDriver at https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html. WDYT?

Yes! As the last fallback, instead of using the non-versioned `LATEST_RELEASE`, let's grab the chromedriver build archive from https://commondatastorage.googleapis.com/chromium-browser-snapshots/{PLATFORM}/LAST_CHANGE , because in this code path we are most likely running the latest dev version of Chrome and the ToT build is more likely to work than the latest release.

Hexcles committed:
Fall back to tip-of-tree Chromium build

Also support "Chromium" (in addition to "Google Chrome") when parsing
version strings.

Hexcles commented:
I've directly modified the PR to use the tip-of-tree ChromeDriver from Chromium build archive as the last fallback. This should PR should also fix the massive failures of ChromeDriver in `webdriver/`, as the necessary patch is included in the latest ChromeDriver.

Since I'm now a co-author of the PR, I won't review it myself. @gsnedders can you take a look?

foolip commented:
Ping @gsnedders for review. Or perhaps @LukeZielinski can take a look?

@Hexcles can you fix the CI failures?

Hexcles committed:
Various misc style improvements

* Factor URL construction out
* Fix logging (mozlog does not support lazy format strings)
* Fix whitespaces (two blank lines between classes, one space after #)

Hexcles commented:
Thank you to @JohnChen0 for the original patchset and to @jgraham for reviewing. This now merged and we should have much better/accurate Chrome results now.
